File: src/ObjectManager/Object.Tes/FilePacks/Records/345-DIAL.Dialog Topic.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class DIALRecord : Record
    {
        internal static DIALRecord LastRecord;

        public enum DIALType : byte
        {
            RegularTopic = 0, Voice, Greeting, Persuasion, Journal
        }

        public override string ToString() => $"DIAL: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public STRVField FULL; // Dialogue Name
        public BYTEField DATA; // Dialogue Type
        public List<FMIDField<QUSTRecord>> QSTIs; // Quests (optional)
        public List<INFORecord> INFOs = new List<INFORecord>(); // Info Records

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); LastRecord = this; return true;
                case "FULL": FULL = new STRVField(r, dataSize); return true;
                case "DATA": DATA = new BYTEField(r, dataSize); return true;
                case "QSTI":
                case "QSTR": if (QSTIs == null) QSTIs = new List<FMIDField<QUSTRecord>>(); QSTIs.Add(new FMIDField<QUSTRecord>(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-DIAL.Dialog Topic.swift
    //
//  DIALRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class DIALRecord: Record {
    internal static var lastRecord: DIALRecord!

    public enum DIALType: UInt8 {
        case regularTopic = 0, voice, greeting, persuasion, journal
    }

    public override var description: String { return "DIAL: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var FULL: STRVField! // Dialogue Name
    public var DATA: BYTEField! // Dialogue Type
    public var QSTIs: [FMIDField<QUSTRecord>]? = nil // Quests (optional)
    public var INFOs = [INFORecord]() // Info Records

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize); DIALRecord.lastRecord = self
        case "FULL": FULL = r.readSTRV(dataSize)
        case "DATA": DATA = r.readT(dataSize)
        case "QSTI",
             "QSTR": if QSTIs == nil { QSTIs = [FMIDField<QUSTRecord>]() }; QSTIs?.append(FMIDField<QUSTRecord>(r, dataSize))
        default: return false
        }
        return true
    }
}
